### ISSUE TITLE
Something like theory attribute at class level?

### ISSUE THREAD

**Issue by onkarjit** [xunit/xunit#2050]
Does xUnit support something like theory attribute but then at the class level which should automatically apply to all the test methods in the test class?

My system has 4 different role. Each test method should be tested for each role. I accomplish this by using the Theory attribute above the test method. 

Right now I have to decorate each test method individually with the Theory attribute. It would be much better if I could just decorate my classes with the Theory attribute and if a method needs to override the Theory attribute at the class level it should provide its own Theory attribute.

**Comment by ChrisMissal** [comment]
@onkarjit Sounds like you might be able to use a [ClassFixture](https://xunit.net/docs/shared-context.html#class-fixture).

**Comment by onkarjit** [comment]
> @onkarjit Sounds like you might be able to use a [ClassFixture](https://xunit.net/docs/shared-context.html#class-fixture).

If I understand ClassFixture correctly it won't execute the same test method multiple times for different roles that I have.

**Comment by psxvoid** [comment]
> Sounds like you might be able to use a [ClassFixture](https://xunit.net/docs/shared-context.html#class-fixture).

For my use-case it's not possible to use a class fixture and I'm definitely looking forward into adding `Theory` on a class-level. Existing class fixtures are created once per test class. I want be able to inject a fixture into a test class, and pass `InlineData` into the constructor, so each method in the test class can contain a single assert per method.

For example, instead of this:
```cs
public class InvalidArguments : IClassFixture<MyFixture>
{
    readonly MyFixture fixture;

    public InvalidArguments(MyFixture fixture)
    {
        this.fixture = fixture;
    }

    [Theory]
    [InlineData(1L)]
    [InlineData(2L)]
    public async Task InvalidDelta_ShouldThrowArgumentException(long delta)
    {
        this.fixture.Delta = delta;
        this.fixture.DeltaUnit = unit;

        ArgumentException ex = await Assert.ThrowsAsync<ArgumentException>(() => this.fixture.Act());

        Assert.Equal("My message.", ex.Message);
        this.fixture.Dependency.Verify(x => x.MyMethod(), Times.Once);
    }
```

I want to have this:
```cs
public class InvalidArguments : IClassFixture<MyFixture>
{
    readonly MyFixture fixture;

    [Theory]
    [InlineData(1L)]
    [InlineData(2L)]
    public InvalidArguments(MyFixture fixture)
    {
        this.fixture = fixture;
        this.fixture.Delta = delta;
        this.fixture.DeltaUnit = unit;
        
        // not possible to call async method from the constructor
        // a workaround is probably needed, e.g. another attribute for a fixture
        await this.fixture.Act();
    }

    public async Task InvalidDelta_ShouldThrowArgumentException()
    {
        Assert.Equal("My message.", ex.Message);
    }
    
    public async Task InvalidDelta_ShouldStillCallMyMethod()
    {
        this.fixture.Dependency.Verify(x => x.MyMethod(), Times.Once);
    }
```

In other words, what is missing:
1. Being able to have `async` setup for a `IClassFixture<T>`
2. Being able to call a method ONCE per each theory test

**Comment by bradwilson** [comment]
I don't foresee us adding something like this, because it dramatically complicates the way constructor arguments are resolved: it's hard to code, and it's hard to reason about.

> // not possible to call async method from the constructor

We offer `IAsyncLifetime` for that.

**Comment by agray** [comment]
> something like theory attribute but then at the class level which should automatically apply to all the test methods in the test class?

As I understand your requirement, I would say yes. I use IClassFixture in Saucery:

https://github.com/search?q=repo%3ASauceforge%2FSaucery%20IClassFixture&type=code

 https://sauceforge.github.io/

**Comment by psxvoid** [comment]
> We offer IAsyncLifetime for that.

I've just quickly setup a test project with a fixture implementing `IAsyncLifetime` and seems like `TestContext.Current.Test as IXunitTest` is `null`:
```cs
    public class MyFixture : IAsyncLifetime
    {
        public Sut sut;

        public Mock<IDependency> Dependency { get; } = new();

        public MyFixture()
        {
            Dependency = new();
            sut = new Sut(Dependency.Object);
        }

        public ValueTask DisposeAsync()
        {
            return ValueTask.CompletedTask;
        }

        public async ValueTask InitializeAsync()
        {
            var context = TestContext.Current.Test as IXunitTest;
           
            // null reference exception here because context is `null`
            var arguments = (IEnumerable<object>)context.TestMethodArguments;
            var delta = arguments.First() as long?;

            await this.sut.MyMethod(delta!.Value);
        }
    }
```

Are there any ways to make `TestContext.Current.Test` available in `...: IClassFixture<T>, IAsyncLifetime` in `InitializeAsync` implementation? IMHO, ideally it would be cool to pass theory arguments into `InitializeAsync`.

**Comment by bradwilson** [comment]
> Update: Managed to achieve what I want by moving `IAsyncLifetime` into the test class, instead of fixture.

Correct, there is no active test when a fixture is being created (because its lifetime is larger than just one test), so `TestContext.Current.Test` would be null during fixture creation/cleanup.

**Comment by bradwilson** [comment]
I am closing old issues today. This issue is one I consider "not planned", as after reconsideration, I agree with my initial thought that this is not a feature we're planning to add.
